v2.1

rstoneback opened the pull request:
# Description
Release our accumulated work since the last release. Thanks for all of the hard work!

Addresses # (issue)

The numerous changes are listed in the changelog.

## Type of change

Please delete options that are not relevant.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Travis CI and local tests by various users throughout developmeny.

# Checklist:

- [ ] Make sure you are merging into the ``develop`` (not ``master``) branch
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] Add a note to ``CHANGELOG.md``, summarizing the changes

rstoneback commented:
Take this out for a spin. See how it works for a few days.

jklenzing commented:
Going to be a while before I can get to this. @JonathonMSmith could you take a look at develop in the meantime?

aburrell commented on pysat/_instrument.py:
```diff
@@ -2050,6 +2141,16 @@ def to_netcdf4(self, fname=None, base_instrument=None, epoch_name='Epoch',
+        # check if there are multiple variables with same characters
+        # but with different case
+        variables = self.variables
+        variables = [var.lower() for var in variables]
```
Why not combine these two: `variables = [var.lower() for var in self.variables]`?  Alternatively: `self.variables = [var.lower() for var in self.variables]` and using `self.variables` elsewhere.

rstoneback replied:
self.variables is a property that returns the names as in the underlying pandas or xarray object. I've made the code more efficient, as suggested.

aburrell commented on pysat/_instrument.py:
```diff
@@ -2257,15 +2366,15 @@ def to_netcdf4(self, fname=None, base_instrument=None, epoch_name='Epoch',
                                 # attach any meta data
                                 try:
-                                    new_dict = export_meta[key+'_'+col]
+                                    new_dict = export_meta[case_key + '_'+col]
```
Spacing not consistent here.

aburrell commented on pysat/_meta.py:
```diff
@@ -270,12 +270,18 @@ def drop(self, names):
         drop_names = []
         for name in current_names:
-            if name not in keep_names:
+            if name not in keep_:
```
I thought it was convention to only use a single trailing underscore to avoid a name conflict with an existing python variable/function.  That doesn't seem to be the case with `keep`.  Am I missing something?

aburrell commented on pysat/_meta.py:
```diff
@@ -517,26 +523,50 @@ def __getitem__(self, key):
+                return [func(name_) for name_ in names[name]]
+            else:
+                # assume iterable
+                return [func(name_) for name_ in name]
```
Same thought as above here.  Why not use a normal string like `nn`?

aburrell commented on pysat/instruments/sw_kp.py:
```diff
@@ -284,26 +284,29 @@ def download(date_array, tag, sat_id, data_path, user=None, password=None):
+                except ftplib.error_perm as exception:
+
+                    if str(exception.args[0]).split(" ", 1)[0] != '550':
+                        raise
```
Should we specify what we're raising here?

rstoneback replied:
The bare raise preserves the stack trace back to the issue with ftp.

https://stackoverflow.com/questions/2052390/manually-raising-throwing-an-exception-in-python

aburrell replied:
Good to know.  Perhaps adding that as a comment for future developers would be useful and help people adding new instruments.

rstoneback replied:
Comment has been added

aburrell commented on pysat/tests/test_ssnl_plot.py:
```diff
@@ -71,3 +71,27 @@ def test_multiple_scatterplots(self):
+    def test_deprecation_warning_scatterplot(self):
+        """Test if scatterplot in ssnl is deprecated"""
+
+        with warnings.catch_warnings(record=True) as w:
```
Should avoid one-character variable for non-iterators.  Recommend changing `w` to `war`.

aburrell commented on pysat/tests/test_utils.py:
```diff
@@ -38,6 +39,23 @@ def remove_files(inst):
+    """Test if computational form in utils is deprecated"""
+
+    import pandas as pds
+    import warnings
```
These should be moved to the top level of the file to avoid dual imports.

aburrell commented on pysat/tests/test_utils.py:
```diff
@@ -260,3 +427,50 @@ def test_write_and_read_netcdf4_default_format_higher_order_w_zlib(self):
+        try:
+            assert self.testInst.bespoke # should raise
+        except AttributeError:
+            pass
```
What is this testing?  What should it be raising?  Can this be tested using `raises`?

rstoneback replied:
The specific lines are a pre-test to ensure a proper setup, that the attribute testInst.bespoke does not exist. The next lines set testInst.meta.bespoke and it is transferred to inst using a method. Then we test that testInst.bespoke is True. If so, we know it came from meta.

aburrell commented on pysat/tests/test_utils_coords.py:
```diff
@@ -108,65 +109,21 @@ def test_bad_lon_name_calc_solar_local_time(self):
-            scale = coords.scale_units(out_unit, "m")
+        warnings.simplefilter("always")
+        scale1 = pysat.utils.scale_units("happy", "happy")
+        with warnings.catch_warnings(record=True) as w:
```
Again, recommend avoiding single letter variables that are not indices.

aburrell commented on pysat/tests/test_utils_stats.py:
```diff
@@ -57,3 +56,55 @@ def test_circstd_nan(self):
+    def test_deprecation_warning_median1d(self):
+        """Test if median1D in stats is deprecated"""
+
+        with warnings.catch_warnings(record=True) as w:
```
Single letter warning variable to be fixed.

aburrell commented on setup.py:
```diff
@@ -7,39 +7,36 @@
     # read the docs doesn't do Fortran
     # remove pysatCDF through h5py
-    install_requires = install_requires[:-8]
+    install_requires = install_requires[:-4]
```
Just a thought, but would it be better to have a `rtd_install_requires` defined in each or perhaps add a comment where `install_requires` is defined to ensure future developers don't overlook this when adding new requirements?

aburrell commented on pysat/tests/test_meta.py:
```diff
@@ -337,6 +337,17 @@ def test_basic_pops_w_bad_key(self):
+        try:
+            self.meta[1]
+        except NotImplementedError:
+            pass
```
Again, would these be better tested using `raises`?

rstoneback commented:
Thanks for the feedback!

jklenzing commented on CHANGELOG.md:
```diff
@@ -2,7 +2,57 @@
+   - Fixed a bug when trying to combine empty kp lists
+   - Updated travis.yml to work with python 2.7.15
+   - Unit tests reload pysat_testing_xarray for xarray tests
+   - Updated setup.py to not overwrite defauly `open` command from `codecs`
```
Spelling of default

jklenzing commented on CHANGELOG.md:
```diff
@@ -2,7 +2,57 @@
+   - Updated travis.yml to work under pysat organization
+   - Added missing requirements (matplotlib, netCDF4)
+   - Fixed a bug when trying to combine empty kp lists
+   - Updated travis.yml to work with python 2.7.15
```
... 2.7.15 and beyond

jklenzing commented on pysat/__init__.py:
```diff
@@ -60,16 +60,16 @@
+        data_dir = '/home/travis/build/pysatData'
     else:
-        data_dir = '/home/travis/build/rstoneback/pysatData'
+        data_dir = ''
```
Won’t this just dump all the data to the current directory?

rstoneback replied:
No, I put a check in the files class that ensures data_dir is not ''. Line 152 in _files.py.

jklenzing commented on pysat/tests/test_model_utils.py:
```diff
@@ -0,0 +1,101 @@
+    def test_satellite_view_through_model_deprecation(self):
+        """Test if satellite_view_through_model is deprecated"""
+
+        with warnings.catch_warnings(record=True) as w:
```
More warnings to clean up

jklenzing replied:
Note this occurs throughout the TestDeprecation class in the tests for modelutils and ssnl

jklenzing commented on pysat/tests/test_utils_time.py:
```diff
@@ -165,3 +165,23 @@ def test_create_datetime_index_wo_month_day_uts():
+    start = pds.datetime(2012, 2, 28)
+    stop = pds.datetime(2012, 3, 1)
+    warnings.simplefilter("always")
+    with warnings.catch_warnings(record=True) as w1:
```
War1 and war2 for consistency

rstoneback commented:
Thanks @aburrell and @jklenzing for the comments on a big pull, and thanks for all the commits as well! I believe I have addressed all comments.

I have a question, probably for Jeff since he has recently worked on DeprecationWarnings. I keep getting a, 
-c:1: DeprecationWarning: Strict times will eventually be enforced upon all instruments. (strict_time_flag)
on every load.

rstoneback commented:
I think I'll just move the warning to instantiation.

jklenzing commented:
Looks like this got caught in my find and replace all to add “stacklevel=2” to the deprecation warnings. This is needed for testing of the other warnings, but may not be needed here.

jklenzing commented on pysat/tests/test_ssnl_occur_prob.py:
```diff
@@ -107,3 +107,74 @@ def test_occur_prob_by_orbit_3D_w_bins(self):
+    def test_deprecation_warning_daily_2D(self):
+        """Test if occur_prob.daily2D is deprecated"""
+
+        with warnings.catch_warnings(record=True) as w:
```
W => war
Sorry, more files to fix throughout

rstoneback commented:
As written we appear to be up-to-date with standards. From
https://docs.python.org/3.7/library/warnings.html
(I checked 2.7 as well) the stacklevel=2 just resets where it looks like the warning originates from.

def deprecation(message):
    warnings.warn(message, DeprecationWarning, stacklevel=2)

jklenzing commented:
> Looks like this got caught in my find and replace all to add “stacklevel=2” to the deprecation warnings. This is needed for testing of the other warnings, but may not be needed here.

Note that this was added to force deprecation warnings in python 2.7, so it may be needed here too.

jklenzing commented:
I remember something in the python 2.7 documentation about deprecation warnings only being needed for developers, not end users, so they shut it off by default.

rstoneback commented:
OK, you are correct that changing the stacklevel does change behavior in 2.7. I don't see it anymore if I reduce the level or remove it entirely.

jklenzing commented:
I think moving to instantiation is the right path

rstoneback commented:
Cool. I've moved the check and, for the moment, I've retained the stacklevel=2 for consistency across python 2 and 3. Also updated a few more 'as w:' areas.

rstoneback commented:
I've spent a bit more time thinking about the strict_time_flag deprecationwarning. While I can see python's larger point of not forcing users to see a DeprecationWarning, in this case I'm trying to provide scientists a heads up about data handling. pysat is most definitely cleaner without the warning, but if nobody ever sees it.... what's the point?

I could go either way on this. Please let me know if you have any feelings on the topic.

The pull is otherwise ready to go, pending additional comments. Cheers!

jklenzing commented:
I’m leaning the same way. Plus it does appear in python 3.7 regardless of stacklevel, so clearly they changed their minds as well.
